source/common/ssl/context_manager_impl.cc

```diff
@@ -22,32 +58,94 @@ void ContextManagerImpl::releaseContext(Context* context) {
 
 ClientContextPtr ContextManagerImpl::createSslClientContext(Stats::Scope& scope,
                                                             ClientContextConfig& config) {
-
   ClientContextPtr context(new ClientContextImpl(*this, scope, config));
-  std::unique_lock<std::mutex> lock(contexts_lock_);
+  std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
   return context;
 }
 
-ServerContextPtr ContextManagerImpl::createSslServerContext(Stats::Scope& scope,
-                                                            ServerContextConfig& config) {
-  ServerContextPtr context(new ServerContextImpl(*this, scope, config, runtime_));
-  std::unique_lock<std::mutex> lock(contexts_lock_);
+bool ContextManagerImpl::isWildcardServerName(const std::string& name) {
+  return name.size() > 2 && name[0] == '*' && name[1] == '.';
+}
+
+ServerContextPtr ContextManagerImpl::createSslServerContext(
+    const std::string& listener_name, const std::vector<std::string>& server_names,
+    Stats::Scope& scope, ServerContextConfig& config, bool skip_context_update) {
+  ServerContextPtr context(new ServerContextImpl(*this, listener_name, server_names, scope, config,
+                                                 skip_context_update, runtime_));
+  std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
+
+  // Save mappings.
+  if (server_names.empty()) {
+    map_exact_[listener_name][EMPTY_STRING] = context.get();
+  } else {
+    for (const auto& name : server_names) {
+      if (isWildcardServerName(name)) {
+        map_wildcard_[listener_name][name] = context.get();
+      } else {
+        map_exact_[listener_name][name] = context.get();
+      }
+    }
+  }
+
   return context;
 }
 
-size_t ContextManagerImpl::daysUntilFirstCertExpires() {
-  std::unique_lock<std::mutex> lock(contexts_lock_);
+ServerContext* ContextManagerImpl::findSslServerContext(const std::string& listener_name,
+                                                        const std::string& server_name) const {
+  // Find Ssl::ServerContext to use. The algorithm for "www.example.com" is as follows:
+  // 1. Try exact match on domain, i.e. "www.example.com"
+  // 2. Try exact match on wildcard, i.e. "*.example.com"
+  // 3. Try "no SNI" match, i.e. ""
+  // 4. Return no context and reject connection.
+
+  // TODO(PiotrSikora): make this lockless.
+  std::shared_lock<std::shared_timed_mutex> lock(contexts_lock_);
+
+  // TODO(PiotrSikora): refactor and combine code with RouteMatcher::findVirtualHost().
+  auto listener_map_exact = map_exact_.find(listener_name);
```

Comment:
nit: I think most of the vars in this function can be const (including the auto ones).

Reply:
Done.